Implement IEnricher interface

matt0bi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines <https://github.com/spotify/gordon-gcp/blob/master/CONTRIBUTING.rst>
2. If the PR is unfinished, please prefix the subject line with [WIP], [DRAFT], or [RFC].
-->

**What this PR does / why we need it**:

This PR implements the IEnricherClient interface. It does not implement the Builder or the Spotify-specific parts of the code - those will be left for future PRs. 

**Which issue(s) this PR fixes** 
<!-- optional; in `fixes #<issue number>, fixes #<issue_number>, ...` format, will close the issue(s) when PR gets merged: -->
N/A

**Special notes for your reviewer**:

This PR is rebased off #9 because it depends on certain parts of that PR to function, which is why this PR is in "RFC" state. If you'd like to give me feedback before #9 is merged, check out commit 5525260.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

@spotify/alf

matt0bi commented on tox.ini:
```diff
@@ -10,7 +10,7 @@ setenv =
 deps = -rdev-requirements.txt
 commands =
     /usr/bin/find . -name "*.pyc" -delete
-    flake8 {posargs}
```
I removed this part because passing in any arguments to tox meant that they would either get passed to flake8 _or_  py.test. Since they have different arguments, this always led to one of them failing.

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+    return {
+        'keyfile': fake_keyfile,
+        'scopes': [],
+        'dns_zone': 'example.com',
```
trailing dot? `example.com.`?

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+
+
+@pytest.fixture
+def event_msg(mocker):
```
Maybe this could be moved to `conftest.py` since I also have a fixture like this in the event consumer plugin.

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+
+
+@pytest.fixture
+def channel_pair():
```
oo this too, to move into `conftest.py`

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+    get_json_mock, get_json_coro = get_mock_coro()
+    http_client = mocker.Mock()
+    mocker.patch.object(http_client, 'get_json', get_json_coro)
+    mocker.patch.object(http_client, '_get_json', get_json_mock)
```
Hmm this is a little confusing. Maybe instead of `_get_json`, `get_json.mock` or `get_json_mock` if not possible? I actually looked up if there was a `_get_json` method in the http client, lol.

matt0bi replied:
Thanks for pointing it out. Changed it to `get_json_mock`.

> I actually looked up if there was a _get_json method in the http client, lol.

GOT YOU! :grin:

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+def mock_http_client(mocker, get_mock_coro):
+    get_json_mock, get_json_coro = get_mock_coro()
+    http_client = mocker.Mock()
+    mocker.patch.object(http_client, 'get_json', get_json_coro)
```
What's the diff between using `mocker.patch...` and `monkeypatch.setattr`? Is it just one less fixture to pass in?

matt0bi replied:
Comparing [monkeypatch](https://docs.pytest.org/en/latest/reference.html#_pytest.monkeypatch.MonkeyPatch.setattr) and [mocker.patch.object](https://docs.python.org/3/library/unittest.mock.html#patch-object) there doesn't seem to be a difference in this case except the "one less fixture". I think I used it because I was using `mocker` everywhere and forgot about `monkeypatch`.

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+
+    assert 'publish' == event_message.phase
+    expected_history_msg = 'Enriched msg 1 with resourceRecords'
+    assert expected_history_msg == event_message.history_log[0]['message']
```
maybe also assert `'enrich' == event_message.history_log[0]['plugin']`?

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+
+    await gce_enricher.process(event_msg)
+
+    event_message = await success_chnl.get()
```
Naming is a bit confusing with `event_msg` fixture. Maybe something like `ret_event_msg`? just something to clearly differentiate.

matt0bi replied:
ಠಿ_ಠಿ, good point!

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+@pytest.mark.asyncio
+async def test_process_event_msg(config, event_msg, channel_pair,
+                                 mock_async_sleep, mock_http_client, caplog,
+                                 network_data, sleep_calls, logs_logged):
```
docstring?

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+async def test_process_event_msg_failures(config, event_msg, channel_pair,
+                                          mock_async_sleep, mock_http_client,
+                                          caplog, response, sleep_calls,
+                                          logs_logged):
```
docstring?

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+    assert event_msg == event_message
+    assert sleep_calls == mock_async_sleep.call_count
+    assert logs_logged == len(caplog.records)
+    assert 0 == len(event_message.history_log)
```
Hmm maybe there should be a log entry for the message history about this. What do you think?

econchick commented on tests/unit/plugins/test_enricher.py:
```diff
@@ -19,18 +19,132 @@
+    await gce_enricher.process(event_msg)
+
+    event_message = await error_chnl.get()
+
```
What do you think of adding an `assert success_chnl.empty()`?

econchick commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
+            return
+        logging.debug(f'Created resource record {a_record}')
+        event_message.data['resourceRecords'].append(a_record)
+        msg = f'Enriched msg {event_message.msg_id} with resourceRecords'
```
Hmm - I wonder if could this history log message be more helpful. You shouldn't need the message's ID since the history log attached to the object itself. I wonder if it'd be helpful to include something like `Enriched msg with {len(added_a_records)} A records.` (and maybe so we can extend it to `{len(added_cname_records)} CNAME records` and TXT etc.)

Also - would you mind adding a period at the end of your (event) log messages? I think I tried to do that consistently w/ my patch. Total nit, but I feel like I'd go crazy with inconsistency.

econchick commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
+            logging.warning(msg)
+            await self.error_channel.put(event_message)
+            return
+        logging.debug(f'Created resource record {a_record}')
```
When you rebase my stuff, could you use the message logger so we get the message ID?

matt0bi commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
+        dns_zone = self.config['dns_zone']
+        hostname = data['resourceName'].split('/')[-1]
+
+        name = '.'.join([hostname, dns_zone])
```
Good catch - I'll add this to the builder's validation flow in #12 .

matt0bi commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
+        return {
+            'name': name,
+            'rrdatas': [external_ip],
+            'type': 'A'
```
Hm, I set it up so that it's part of the enricher config/workflow - could you take a look and see if it makes sense this way? This would keep event message changes to just one plugin. What do you think?

matt0bi commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
+        await self.update_phase(event_message, 'publish')
+        await self.success_channel.put(event_message)
+
+    async def update_phase(self, event_message, phase):
```
> Maybe this should live on the message itself? since it's acting on the message?

Talked about about this IRL: this will be moved to the msg when working on core routing logic.

matt0bi commented:
Hey @econchick, I've addressed your feedback, could PTAL? Also, some of your feedback also applies to #12, so I'll go over that on next occasion and apply it there.

econchick commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +57,96 @@ class GCEEnricher:
+            return
+        msg_logger.debug(f'Created resource record {a_record}')
+        event_message.data['resourceRecords'].append(a_record)
+        added_a_records = event_message.data['resourceRecords']
```
So, hm, how would someone (us, too) extend this to process more record types? It doesn't seem like one would be able to without rewriting all of `process`. Maybe, `process` could iterate over some registration of record types that are mapped to their respective methods. Something like (I have no idea if this would work, and this is just the first approach I thought of, there may be better ones):

```python
# original
class GCEEnricher:
    RECORD_HANDLERS = {
        'A': '_create_A_rrecord',  # string name of method
    }
    async def process(self, msg):
        # prep stuff
        for rtype, method in RECORD_HANDLERS.items():
            records = await getattr(self, method)(data, msg_logger)
        # other stuff

# extended
class MyEnricher(GCEEnricher):
    RECORD_HANDLERS.update({ 
        'CNAME': '_create_cname_rrecord',
    }
```

matt0bi replied:
Thanks for pointing this out. I went with the approach we're using in the Janitor code eg. overwritable methods. Users that want more records can override the `_create_rrecords` method to call other, new methods for creating specific records. I'm not 100% sure this is the most comfortable way, but it seems the simplest.

econchick commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +57,96 @@ class GCEEnricher:
+
+    async def _poll_for_ext_ip(self, resource_name, msg_logger):
+        exception = None
+        backoff = 2
```
Hmm we should put backoff in the http client eventually 🤔

matt0bi replied:
Created a ticket off this.

matt0bi commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +57,110 @@ class GCEEnricher:
+                msg_logger.debug(f'Attempt {attempt}: fetching {base_url}')
+                instance_data = await self._http_client.get_json(base_url)
+
+                a_confs = instance_data['networkInterfaces'][0]['accessConfigs']
```
This piece of code simply checks for the existence of a nested key. I think assigning this variable is the lesser evil compared to two other alternatives I thought of: adding an `if` block, which adds yet one more level of indentation, or pulling this out into its own tiny function. Any thoughts/advice?

econchick replied:
Another approach could be an `assert` statement:

```python
assert instance_data['networkInterfaces'][0]['accessConfigs'][0]['natIP'], 'some error message'
```

(maybe shorten that dict lookup somehow)

I don't necessarily have strong opinions either way.

matt0bi replied:
Ah, thanks. Makes the intent more explicit :+1:

matt0bi commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +57,96 @@ class GCEEnricher:
+            event_message (.GEventMessage): message requiring
+                additional information.
+        """
+        msg_logger = _utils.GEventMessageLogger(
```
I think there is - do you want to talk about this IRL today?

econchick commented on src/gordon_gcp/plugins/enricher.py:
```diff
@@ -49,11 +57,110 @@ class GCEEnricher:
+        old_phase = event_message.phase
+        new_phase = phase or self.phase
+        event_message.phase = new_phase
+        msg = f'Updated phase from "{old_phase}" to "{phase}".'
```
Hm shouldn't this be `new_phase` instead of `phase`?

matt0bi commented:
Thanks @econchick, I addressed the two pieces of feedback! Will merge this in either today or Monday.
